src/jit/flowgraph.cpp

```diff
@@ -1877,6 +1877,48 @@ void Compiler::fgComputeReachabilitySets()
 }
 
 /*****************************************************************************
+ *  Updates BBF_GC_SAFE_POINT flag for each block
+ */
+
+void Compiler::fgUpdateSafepointFlags()
+{
+    if (compTailCallUsed)
+    {
+        for (BasicBlock* block = fgFirstBB; block != nullptr; block = block->bbNext)
+        {
+            block->bbFlags &= ~BBF_GC_SAFE_POINT;
+            for (GenTreeStmt* stmt = block->firstStmt(); stmt != nullptr; stmt = stmt->getNextStmt())
+            {
+                for (GenTree* node = stmt->gtStmtExpr; node != nullptr; node = node->gtNext)
+                {
+                    if (node->IsCall())
+                    {
+                        GenTreeCall* call = node->AsCall();
+                        if (call->IsTailCall())
+                        {
+                            if (call->IsTailCallViaHelper())
+                            {
+                                block->bbFlags |= BBF_GC_SAFE_POINT;
+                            }
+                        }
+                        else
+                        {
+                            block->bbFlags |= BBF_GC_SAFE_POINT;
+                        }
+                    }
```

Comment:
You should instead pull out lines 1896- 1907 into a new method:
bool fgIsGcSafePoint (or something similar)
I would make it return true if the call is a GC Safe point and then set the BBF_GC_SAFE_POINT here and in morph.cpp

Then you should call the new method from here and from morph.cpp line 8296

Note that your code is setting  BBF_GC_SAFE_POINT for all helper calls and that isn't correct.
In particular calls to the GC Write Barrier helpers are not GC safe points.
Conversely with our existing code it is also not exactly correct to say that all helper calls are not GC safe points, 
but is it conservatively safe to do so.

We should eventually add a new method to HelperCallProperties that returns true if the helper call is a GC safe point.   However you probably shouldn't address that with this check in.  Opening a new GitHib issue to do so in a follow on checkin would probably be best.

here is the code from Morph that your new method should closely match:

```
#ifdef _TARGET_AMD64_
    // Amd64 note: If this is a fast tail call then don't count it as a call
    // since we don't insert GC-polls but instead make the method fully GC
    // interruptible.
    if (!call->IsFastTailCall())
#endif
    {
        if (call->gtCallType == CT_INDIRECT)
        {
            compCurBB->bbFlags |= BBF_GC_SAFE_POINT;
        }
        else if (call->gtCallType == CT_USER_FUNC)
        {
            if ((call->gtCallMoreFlags & GTF_CALL_M_NOGCCHECK) == 0)
            {
                compCurBB->bbFlags |= BBF_GC_SAFE_POINT;
            }
        }
        // otherwise we have a CT_HELPER
    }
...